### PR TITLE
add thread/messages API

### PR DESCRIPTION
**Pull request by PetrDlouhy**
Add API endpoint for public threads and messages.

This is proposal, please comment with your suggestions.

### CODE REVIEW

File: app/api/route/message_api.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      optional :thread, type: Integer, desc: 'ID of thread'
+      optional :order_by, type: Symbol, values:%i( created_at  ), desc: 'Order of returned issues.'
+      optional :order, type: Symbol, values:%i(asc desc), default: :asc
+      optional :submit_external, type: String, desc: 'Filter by external service submission mark'
+    end
+
+    helpers do
+      def post_or_get_messages
+        scope = Message.joins(:thread).includes(:created_by)
+        scope = scope.where('message_threads.privacy' => 'public')
```
Comment (nikolai-b) [inline app/api/route/message_api.rb]: I'd prefer
```ruby
scope = scope.merge(MessageThread.is_public)
```
also we need `.approved` or we might show spam messages

File: app/api/route/message_api.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        scope = scope.where(censored_at: nil)
+        scope = scope.order(params[:order_by] => params[:order]) if params[:order_by]
+        scope = scope.where(thread_id: params[:thread]) if params[:thread]
+        scope = scope.where('message_threads.submit_external' => params[:submit_external]) if params[:submit_external]
```
Comment (nikolai-b) [inline app/api/route/message_api.rb]: There is no `submit_external` in the codebase.  Is it something you've added elsewhere?

File: app/api/route/message_api.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
+module Route
+  class MessageApi < Base
+    paginate paginate_settings
+
+    params do
+      optional :thread, type: Integer, desc: 'ID of thread'
```
Comment (nikolai-b) [inline app/api/route/message_api.rb]: I'd prefer if it was called `:thread_id`

File: app/api/route/thread_api.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
+module Route
+  class ThreadApi < Base
+    paginate paginate_settings
+
+    params do
+      optional :group, type: String, desc: 'Return only issues from area of group given by its short name, e.g. "london"'
+      optional :issue, type: Integer, desc: 'ID of issue'
```
Comment (nikolai-b) [inline app/api/route/thread_api.rb]: `:issue_id`?

File: app/api/route/message_api.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    helpers do
+      def post_or_get_messages
+        scope = Message.joins(:thread).includes(:created_by)
+        scope = scope.where('message_threads.privacy' => 'public')
+        scope = scope.where(censored_at: nil)
+        scope = scope.order(params[:order_by] => params[:order]) if params[:order_by]
+        scope = scope.where(thread_id: params[:thread]) if params[:thread]
+        scope = scope.where('message_threads.submit_external' => params[:submit_external]) if params[:submit_external]
+        scope = paginate scope
```
Comment (nikolai-b) [inline app/api/route/message_api.rb]: You've not `def as_json` on the `Message` so we will be sending all the fields that exist in the db, like `check_reason` and `public_token`.  Can we trim it to a specific white list please?
e.g.
https://github.com/cyclestreets/cyclescape/blob/staging/app/decorators/library/item_decorator.rb#L20-L28

File: app/api/route/thread_api.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      optional :order_by, type: Symbol, values:%i( created_at  ), desc: 'Order of returned issues.'
+      optional :order, type: Symbol, values:%i(asc desc), default: :asc
+      optional :submit_external, type: String, desc: 'Filter by external service submission mark'
+    end
+
+    helpers do
+      def post_or_get_thread
+        scope = MessageThread.all.includes(:messages)
```
Comment (nikolai-b) [inline app/api/route/thread_api.rb]: or just
```ruby
scope = MessageThread.is_public.approved.includes(:messages)
```

File: app/api/route/message_api.rb (hunk 5 above)
Comment (mvl22) [reply app/api/route/message_api.rb]: +1

**Comment by mvl22** [comment]
Thanks!

How does this handle non-public threads, in terms of the response that is returned?

File: app/api/route/message_api.rb (hunk 3 above)
Comment (PetrDlouhy) [reply app/api/route/message_api.rb]: OK

File: app/api/route/message_api.rb (hunk 1 above)
Comment (PetrDlouhy) [reply app/api/route/message_api.rb]: OK

File: app/api/route/message_api.rb (hunk 2 above)
Comment (PetrDlouhy) [reply app/api/route/message_api.rb]: Yes, I am working on function, where threads could be marked to be send to the municipality.
I will remove this from this PR, it belongs to separate PR with that functionality.

File: app/api/route/thread_api.rb (hunk 4 above)
Comment (PetrDlouhy) [reply app/api/route/thread_api.rb]: OK

File: app/api/route/thread_api.rb (hunk 6 above)
Comment (PetrDlouhy) [reply app/api/route/thread_api.rb]: OK

**Comment by PetrDlouhy** [comment]
@mvl22 non-public threads are just filtered out. I don't know, how to work with authentication on REST API in RoR, so I am leaving this open for further improvements.

**Comment by mvl22** [comment]
> @mvl22 non-public threads are just filtered out

I think that the API ought to behave in the same way as if an unauthenticated public user (in this case an unauthenticated API request, so a public user) had accessed a non-visible page. Currently `/threads/<id>/` gives a login required redirect - so the equivalent for an API would be `403 Forbidden`, which is the nearest equivalent.

I don't think we should prioritise adding token auth via the API at this time, unless there is a need.

**Comment by PetrDlouhy** [comment]
I have updated this PR by the suggestions.
@nikolai-b Is there something similar to `as_json` also for the XML format?

@mvl22 I don't get it. Nothing like `/api/threads/<id>/` works, so there is no way of getting single thread through API. What I can do is getting non-public threads in format without the additional information like
```
  {
    "id": 755,
    "issue_id": 845,
    "created_by_id": 487,
    "group_id": 1,
    "privacy": "public",
  },
```
in the listing.

**Comment by nikolai-b** [comment]
@PetrDlouhy great, merged into staging

**Pull request merged**